lib/src/type_checker.dart

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:mirrors';
+
+import 'package:analyzer/dart/constant/value.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+
+/// An abstraction around doing static type checking at compile/build time.
+abstract class TypeChecker {
+  const TypeChecker._();
+
+  /// Create a new [TypeChecker] backed by a runtime [type].
+  ///
+  /// This implementation uses `dart:mirrors` (runtime reflection).
+  const factory TypeChecker.fromRuntime(Type type) = _MirrorTypeChecker;
+
+  /// Create a new [TypeChecker] backed by a static [type].
+  const factory TypeChecker.fromStatic(DartType type) = _LibraryTypeChecker;
+
+  /// Create a new [TypeChecker] backed by a library [url].
+  ///
+  /// Example of referring to a `LinkedHashMap` from `dart:collection`:
+  /// ```dart
+  /// const linkedHashMap = const TypeChecker.fromUrl(
+  ///   'dart:collection#LinkedHashMap',
+  /// );
+  /// ```
+  const factory TypeChecker.fromUrl(dynamic url) = _UriTypeChecker;
+
+  /// Returns the first constant annotating [element] that is this type.
+  ///
+  /// Otherwise returns `null`.
+  DartObject annotationOf(Element element) {
+    for (final metadata in element.metadata) {
+      final constant = metadata.computeConstantValue();
+      if (isExactlyType(constant.type)) {
+        return constant;
+      }
+    }
+    return null;
+  }
+
+  /// Returns `true` if representing the exact same class as [element].
+  bool isExactly(Element element);
+
+  /// Returns `true` if representing the exact same type as [staticType].
+  bool isExactlyType(DartType staticType) => isExactly(staticType.element);
+
+  /// Returns `true` if representing a super class of [element].
+  bool isSuperOf(Element element) =>
+      isExactly(element) ||
+      element is ClassElement && element.allSupertypes.any(isExactlyType);
+
+  /// Returns `true` if representing a super type of [staticType].
+  bool isSuperType(DartType staticType) => isSuperOf(staticType.element);
+}
+
+// Checks a static type against another static type;
+class _LibraryTypeChecker extends TypeChecker {
+  final DartType _type;
+
+  const _LibraryTypeChecker(this._type) : super._();
+
+  @override
+  bool isExactly(Element element) =>
+      element is ClassElement && element.type == _type;
+
+  @override
+  bool isSuperOf(Element element) =>
+      isExactly(element) ||
+      element is ClassElement &&
+          element.allSupertypes.any((t) => t.element == _type.element);
```

Comment:
Why is the logic different for checking super types? (here it checks for ClassElement equality where `isExactly` checks for DartType equality)